PG.StarWarsGame.Files.MEG/PG.StarWarsGame.Files.MEG/Services/V1/MegFileProcessService.cs

```diff
@@ -0,0 +1,251 @@
+{
+    /// <summary>
+    ///     Default implementation of <see cref="IMegFileProcessService" /> for
+    ///     <a href="https://modtools.petrolution.net/docs/MegFileFormat">v1 <code>*.MEG</code> files</a>.
+    ///     When requesting the default implementation via an IoC Container or registering via injection, you may pass
+    ///     a file system as argument implementing <see cref="System.IO.Abstractions.IFileSystem" /> and a logger factory
+    ///     implementing <see cref="Microsoft.Extensions.Logging.ILoggerFactory" />
+    /// </summary>
+    [Export(nameof(IMegFileProcessService))]
+    public sealed class MegFileProcessService : IMegFileProcessService
+    {
+        [NotNull] private readonly IFileSystem m_fileSystem;
+        [CanBeNull] private readonly ILogger m_logger;
+
+        public MegFileProcessService([CanBeNull] IFileSystem fileSystem,
+            [CanBeNull] ILoggerFactory loggerFactory = null)
+        {
+            m_fileSystem = fileSystem ?? new FileSystem();
+            m_logger = loggerFactory?.CreateLogger<MegFileProcessService>();
+        }
+
+        public void PackFilesAsMegArchive(string megArchiveName,
+            Dictionary<string, string> packedFileNameToAbsoluteFilePathsMap, string targetDirectory)
+        {
+            if (!StringUtility.HasText(megArchiveName))
+            {
+                throw new ArgumentException(
+                    "The provided argument is null, an empty string or only consists of whitespace.",
+                    nameof(megArchiveName));
+            }
+
+            if (!packedFileNameToAbsoluteFilePathsMap.Any())
+            {
+                throw new ArgumentException(
+                    "The provided argument does not contain any elements.",
+                    nameof(packedFileNameToAbsoluteFilePathsMap));
+            }
+
+            if (!StringUtility.HasText(targetDirectory))
+            {
+                throw new ArgumentException(
+                    "The provided argument is null, an empty string or only consists of whitespace.",
+                    nameof(targetDirectory));
+            }
+
+            string actualName = StringUtility.StripFileExtension(megArchiveName);
+            MegFileHolder megFileHolder = new MegFileHolder(targetDirectory, actualName);
+            foreach ((string key, string value) in packedFileNameToAbsoluteFilePathsMap)
+            {
+                megFileHolder.Content.Add(new MegFileDataEntry(key, value));
+            }
+
+            MegFileBuilder builder = new MegFileBuilder(m_fileSystem);
+            MegFile megFile = builder.FromHolder(megFileHolder, out List<string> filesToStream);
+            string writePath = m_fileSystem.Path.Combine(megFileHolder.FilePath, megFileHolder.FullyQualifiedName);
+            CreateTargetDirectoryIfNotExists(megFileHolder.FilePath);
+            using (BinaryWriter writer = new BinaryWriter(m_fileSystem.FileStream.Create(writePath, FileMode.Create)))
```

Comment:
What about file blocking? Do we accept concurrent write/read.

Reply:
Nope, concurrenty whilst packaging an archive sounds like way too much trouble too even be worth considering.